### PR TITLE
[GEOT-6653] Add InFunction optimization to MemoryFilterOptimizer

### PR DESCRIPTION
**Pull request by fernandor777**
This PR introduces a performance improvement on MemoryFilterOptimizer for in-memory features filter evaluation.

JIRA issue:
https://osgeo-org.atlassian.net/browse/GEOT-6653

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [x] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer.

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):
- [x] There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [x] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by travis-ci after opening this PR)
- [x] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

File: modules/library/render/src/main/java/org/geotools/renderer/lite/MemoryFilterOptimizer.java [hunk 1]
```diff
@@ -111,7 +117,58 @@ public Object visit(Not filter, Object extraData) {
...
+    /** Checks if an Or filter can be replaced by a single IN condition. */
+    private Object inFilterOptimizer(Or filter, Object extraData) {
+        InFunction inFilter = replaceWithInFilter(filter);
+        if (inFilter == null) return super.visit(filter, extraData);
+        return ff.equals(inFilter, ff.literal(true));
+    }
+
+    private InFunction replaceWithInFilter(Or filter) {
+        List<Filter> children = filter.getChildren();
+        if (children.size() == 0
```
Comment (aaime) [inline modules/library/render/src/main/java/org/geotools/renderer/lite/MemoryFilterOptimizer.java:132]: This code does repeated passes over data structures as long as the or children, while a traditional loop could have done everything in a single pass, just keeping track of the first property name found, and the list of literals found in the process. 

Just saying, streams are not always an improvement, had this been in a performance critical section of code, the penalty would have been severe. But it's in a one-off, so it does not matter much, you can leave it as is.

File: modules/library/render/src/test/java/org/geotools/renderer/lite/MemoryFilterOptimizerTest.java [hunk 2]
```diff
@@ -144,4 +148,35 @@ public void testEqualFeatureTypes() throws Exception {
         Mockito.verify(spy, Mockito.times(0)).getAttribute(name);
         Mockito.verify(spy, Mockito.times(1)).getAttribute(0);
     }
+
+    public void testInFunctionOptimizer() throws Exception {
```
Comment (aaime) [inline modules/library/render/src/test/java/org/geotools/renderer/lite/MemoryFilterOptimizerTest.java:152]: This test covers only the happy case, could you write tests for the exception paths, like or with different properties, and with comparisons other than equality?

File: modules/library/render/src/main/java/org/geotools/renderer/lite/MemoryFilterOptimizer.java [hunk 3]
```diff
@@ -111,7 +117,58 @@ public Object visit(Not filter, Object extraData) {
...
+        List<Expression> inParameters = new ArrayList<>(parameters.size() + 1);
+        inParameters.add(parameters.get(0).getKey());
+        inParameters.addAll(
+                parameters.stream().map(p -> p.getValue()).collect(Collectors.toList()));
+        InFunction inFunction = new InFunction();
+        inFunction.setParameters(inParameters);
+        return inFunction;
+    }
+
+    private Pair<PropertyName, Literal> getEqualsParameters(PropertyIsEqualTo equals) {
```
Comment (aaime) [inline modules/library/render/src/main/java/org/geotools/renderer/lite/MemoryFilterOptimizer.java:157]: Nitpicking here, an "in" optimization can be performed over pairs of "Expression, Literal", e.g., ``strLow(myAtt) = 'abc' or strLow(myAtt) = 'def'`` or ``(a + b) = 10 or (a + 10) = 12`` can be both optimized with the in function. The only requirement is that all non literal bits in the comparison are equal. Again, probably not the most common case, up to you if you want to generalize or not.

File: modules/library/render/src/main/java/org/geotools/renderer/lite/MemoryFilterOptimizer.java (hunk 1 above)
Comment (fernandor777) [reply modules/library/render/src/main/java/org/geotools/renderer/lite/MemoryFilterOptimizer.java:132]: Indeed a single iteration over the collection is possible and faster, so I did the fix, thanks!

File: modules/library/render/src/test/java/org/geotools/renderer/lite/MemoryFilterOptimizerTest.java (hunk 2 above)
Comment (fernandor777) [reply modules/library/render/src/test/java/org/geotools/renderer/lite/MemoryFilterOptimizerTest.java:152]: Agreed, added test cases when the In function optimization is not accomplished.

File: modules/library/render/src/main/java/org/geotools/renderer/lite/MemoryFilterOptimizer.java (hunk 3 above)
Comment (fernandor777) [reply modules/library/render/src/main/java/org/geotools/renderer/lite/MemoryFilterOptimizer.java:157]: Nice hint, added expression support.

**Comment by aaime** [comment]
Looking good, merging, thanks!

**Pull request merged**